man/man3/pnet_reverse.3
.TH pnet_reverse 3  "version 1.0.3" "Pnet programmers manual"
.SH NAME
pnet_reverse \-  Reverse direction of all connections
.SH SYNOPSIS
.nf
\fB#include <mpi.h>
#include "pnet.h"

int pnet_reverse(Pnet\fP *\fIpnet\fB);\fP
.fi
.SH DESCRIPTION
Function reverses direction of all connections.

Connection have to be added before by calls of \fBpnet_addconnect\fP. Function must be called before
\fBpnet_setup\fP.
The argument of \fBpnet_reverse\fP is:
.TP
.I pnet
Pointer to initialized Pnet struct 
.SH RETURN VALUE
Upon successful completion \fBPNET_OK\fP is returned.
.SH ERRORS
.TP
.B PNET_ALLOC_ERR
Cannot allocate memory.

.SH AUTHORS

For authors and contributors see AUTHORS file

.SH COPYRIGHT

(C) Potsdam Institute for Climate Impact Research (PIK), see COPYRIGHT file

.SH SEE ALSO
pnet_exchg(3), pnet_init(3), pnet_addconnect(3), pnet_setup(3), pnet_free(3),  pnet_lo(3), pnet_hi(3), pnet_nodes(3)


CHANGELOG.md
# Changelog

All notable changes to this project will be documented in this file.

The format is based on [Keep a Changelog](https://keepachangelog.com/en/1.0.0/),
and this project follows a [GNU standard version numbering](https://www.gnu.org/prep/standards/html_node/Releases.html#index-version-numbers_002c-for-releases)
of `major.minor.patch` with
- `major` for substantial changes to code functionality,
- `minor` for additions of modules and functionality, and
- `patch` for technical changes and bugfixes without extending the functionality of the code.

- if applicable, categories to include in the CHANGELOG.md are:
  - Added
  - Changed
  - Deprecated
  - Removed
  - Fixed
  - Security

## [Unreleased]

## [5.6.20] - 2023-07-13

### Added

- New flags `"fix_co2"` and `"fix_co2_year"` added.

### Changed

- Flag `"istimber"` renamed to `"luc_timber"`.

- Flag `"new_phenology"` renamed to `"gsi_phenology"`.

- Flag `"shuffle_climate"` renamed to `"shuffle_spinup_climate"`.

- The settings for fixed input data have been updated. For climate and N deposition input an interval can be specified from which data is taken after the specified year. In can specified whether random shuffling or cycling is used:

```java
"fix_climate" : false,                /* fix climate after specified year */
"fix_climate_year" : 1901,            /* year after climate is fixed */
"fix_climate_interval" : [1901,1930],
"fix_climate_shuffle" : true,          /* randomly shuffle climate in the interval */
"fix_deposition_with_climate" : false, /* fix N deposition same as climate */
"fix_deposition" : false,              /* fix N deposition after specified year */
"fix_deposition_year" : 1901,          /* year after deposition is fixed */
"fix_deposition_interval" : [1901,1930],
"fix_deposition_shuffle" : true,       /* randomly shuffle depositions in the interval */
"fix_landuse" : false,                 /* fix land use after specfied year */
"fix_landuse_year" : 1901,             /* year after land use is fixed */
"fix_co2" : false,                     /* fix atmospheric CO2  after specfied year */
"fix_co2_year" : 1901,                 /* year after CO2 is fixed */
```

### Removed

- Obsolete flag `"const_climate"` removed.


## [5.6.19] - 2023-07-06

### Added

- Help option `-h` added to `bin2cdf` and `clm2cdf` utility.

- Target `lpjcheck` added to `Makefile` to create only the `lpjcheck` utility.

### Changed

- The version of all man pages has been set to the LPJmL version number, date of last modification removed.

- The different versions of the utilities have been replaced by a reference to the LPJmL version.

### Removed

- All html files have been removed from the repository.

- Obsolete link `lpj` to `lpjml` removed.


## [5.6.18] - 2023-07-04

### Fixed

- Soil array `wpwp` is now correctly initialized with `soilpar->wpwp` for the `"prescribed_soilpar"` setting.


## [5.6.17] - 2023-06-30

### Fixed

- Replaced logical `or` by logical `and` in `survive.c`. This solves issue [#309](https://gitlab.pik-potsdam.de/lpjml/LPJmL_internal/-/issues/309).


## [5.6.16] - 2023-06-30

### Fixed

- Removed excess `pft->nind` in all harvest routines in `harvest_stand.c`.


## [5.6.15] - 2023-06-30

### Added

- Check for matching size of discharge queues read from restart file added to `initriver()` function in `initdrain.c`. If the size differs from the value calculated from the river lengths the queues will be resized and initialized to zero. Error message is additionally printed.


## [5.6.14] - 2023-06-08

### Removed

- Support for the deprecated `*.conf` configuration file format has been removed. Only JSON files are allowed for configuration files and metafiles for input.

- Man pages for the description of the `*.conf` file format removed.

### Changed

- Strings in JSON configuration files can now be longer than 256 characters.


## [5.6.13] - 2023-05-16

### Added

- Missing update of `"harvestc"` and `"harvestn"` output added in `annual_biomass_tree.c`, `annual_woodplantation.c`, `daily_agriculture_tree.c`, and `annual_agriculture_tree.c`.


## [5.6.12] - 2023-05-12

### Added

- Missing agsub litter pool added to `"litc"`, `"litn"`, `"mg_litc"`, `"litc_agr"`, `"mgrass_litc"`, `"mgrass_litn"` outputs and in global flux of `litc`.

- Output of `"litc_ag"` and `"litc_all"` now written to output file.

### Changed

- Item `ag` in struct `Litter` renamed to `agtop`.

- Functions  `litter_ag_grass()`, `litter_ag_nitrogen_tree()`, `litter_ag_sum()`, `litter_ag_sum_quick()`, `litter_ag_tree()`
  renamed to `litter_agtop_grass()`, `litter_agtop_nitrogen_tree()`, `litter_agtop_sum()`, `litter_agtop_sum_quick()`, `litter_agtop_tree()`.

- Index in `outnames` array in `fwriteoutput.c` corrected for `"mgrass_soilc"`, `"mgrass_soiln"`, `"mgrass_litc"`, `"mgrass_litn"`outputs.

## [5.6.11] - 2023-05-11

### Fixed

- Invalid access to pointers `soil->wsat-soil->wpwp` has been replaced by access to layer specific data `soil->wsat[l]-soil->wpwp[l]` in `infil_perc_rain.c` and `infil_perc_irr.c`. See issue [#303](https://gitlab.pik-potsdam.de/lpjml/LPJmL_internal/-/issues/303).


## [5.6.10] - 2023-05-10

### Added

- PFT specific output for biological nitrogen fixation with grid and PFT scaling in `npp_contr_biol_n_fixation.c` and `biologicalNfixation.c`.


### Changed

- Adjusted optimum temperature limits (`"temp_bnf_opt"`) of herbaceous PFTs for NPP controled biological nitrogen fixation in `pft.js`.


## [5.6.9] - 2023-05-10

### Added

- Missing `break` statement added in `harvest_stand.c`. See issue [#302](https://gitlab.pik-potsdam.de/lpjml/LPJmL_internal/-/issues/302)

- Missing deallocation of `decomp_litter_pft` added in `freesoil.c`.

- Missing deallocation of `with_tillage_filename` added in `freeconfig.c`.

### Changed

- If input comes from socket connection the existence of the file is not checked anymore.

- The memcpy arguments in `cpl.h` are now in the right order.

## [5.6.8] - 2023-04-28

### Added


src/pnet/pnet_reverse.c
/**************************************************************************************/
/**                                                                                \n**/
/**                 p  n  e  t  _  r  e  v  e  r  s  e  .  c                       \n**/
/**                                                                                \n**/
/**     MPI-parallelization of networks                                            \n**/
/**                                                                                \n**/
/**     Function reverses the direction of all connections                         \n**/
/**                                                                                \n**/
/**     Implementation is described in:                                            \n**/
/**     W. von Bloh, 2008. Sequential and Parallel Implementation of               \n**/
/**     Networks. In P. beim Graben, C. Zhou, M. Thiel, and J. Kurths              \n**/
/**     (eds.), Lectures in Supercomputational Neuroscience, Dynamics in           \n**/
/**     Complex Brain Networks, Springer, New York, 279-318.                       \n**/
/**                                                                                \n**/
/** (C) Potsdam Institute for Climate Impact Research (PIK), see COPYRIGHT file    \n**/
/** authors, and contributors see AUTHORS file                                     \n**/
/** This file is part of LPJmL and licensed under GNU AGPL Version 3               \n**/
/** or later. See LICENSE file or go to http://www.gnu.org/licenses/               \n**/
/** Contact: https://github.com/PIK-LPJmL/LPJmL                                    \n**/
/**                                                                                \n**/
/**************************************************************************************/

#include <stdio.h>
#include <stdlib.h>
#include <string.h>
#include <errno.h>
#ifdef USE_MPI
#include <mpi.h>
#endif
#include "types.h"
#include "pnet.h"

typedef struct
{
  int index,from;
} data_t;

static int compare(const data_t *a,const data_t *b)
{
  /* compare function to sort data_t items by their integer index */
  return a->index-b->index;
} /* of 'compare' */

int pnet_reverse(Pnet *pnet /**< Pointer to Pnet structure */
                )           /** \return error code        */
{
  int *lo,*hi;
  int i,j,size,slice,rem,task,outsize,rc;
  data_t *in,*out;
#ifdef USE_MPI
  MPI_Datatype type;
#endif
  if(pnet==NULL)
    return PNET_NULL_PTR_ERR;
  slice=pnet->n/pnet->ntask;
  rem=pnet->n % pnet->ntask;
  lo=newvec(int,pnet->ntask);
  if(lo==NULL)
    return PNET_ALLOC_ERR;
  hi=newvec(int,pnet->ntask);
  if(hi==NULL)
  {
    free(lo);
    return PNET_ALLOC_ERR;
  }
   /* calculate lower and upper bounds for all tasks */
  for(i=0;i<pnet->ntask;i++)
  {
    lo[i]=i*slice;
    hi[i]=(i+1)*slice-1;
    if(i<rem)
    {
       lo[i]+=i;
       hi[i]+=i+1;
    }
    else
    {
      lo[i]+=rem;
      hi[i]+=rem;
    }
  }
  for(i=0;i<pnet->ntask;i++)
    pnet->outlen[i]=pnet->inlen[i]=0;
  /* determine total length of connection lists */
  size=0;
  pnet_foreach(pnet,i)
    size+=pnet->connect[i].n;
  in=newvec(data_t,size);
  if(in==NULL)
  {
    free(lo);
    free(hi);
    return PNET_ALLOC_ERR;
  }
  /* concatenate connection lists to array in */
  size=0;
  for(i=pnet->lo;i<=pnet->hi;i++)
    for(j=0;j<pnet->connect[i].n;j++)
    {
      in[size].from=i;
      in[size++].index=pnet->connect[i].index[j];
    }
  /* sort by their index in ascending order */
  qsort(in,size,sizeof(data_t),(int (*)(const void *,const void *))compare);
  task=0;
  /* calculate inlen vector */
  pnet->inlen[task]=0;
  i=0;
  while(i<size)
    /* check, whether in[i].index is within task */
    if(in[i].index<=hi[task])
    {
      /* yes, increase inlen */
      pnet->inlen[task]++;
      i++;
    }
    else
    {
      /* no, move to next task */
      task++;
      pnet->inlen[task]=0;
    }
  /* empty connection lists  */
  for(i=pnet->lo;i<=pnet->hi;i++)
    emptyintlist(pnet->connect+i);
#ifdef USE_MPI
  MPI_Alltoall(pnet->inlen,1,MPI_INT,pnet->outlen,1,MPI_INT,pnet->comm);
#else
  pnet->outlen[0]=pnet->inlen[0];
#endif
  /* calculate displacement vectors */
  pnet->indisp[0]=pnet->outdisp[0]=0;
  for(i=1;i<pnet->ntask;i++)
  {
    pnet->indisp[i]=pnet->inlen[i-1]+pnet->indisp[i-1];
    pnet->outdisp[i]=pnet->outlen[i-1]+pnet->outdisp[i-1];
  }
#ifdef DEBUG
  printf("in %d:",pnet->taskid);
  j=0;
  for(i=0;i<size;i++)
  {
    if(pnet->indisp[j]+pnet->inlen[j]==i)
    {
      printf("|");
      j++;
    }
    printf(" %d,%d",in[i].from,in[i].index);
  }
  printf("\n");
#endif
  /* calculate length of out vector */
  outsize=0;
  for(i=0;i<pnet->ntask;i++)
    outsize+=pnet->outlen[i];
  out=newvec(data_t,outsize);
  if(out==NULL)
  {
    free(lo);
    free(hi);
    free(in);
    return PNET_ALLOC_ERR;
  }
#ifdef USE_MPI
  /* create new MPI datatype for data_t */
  MPI_Type_contiguous(2,MPI_INT,&type);
  MPI_Type_commit(&type);
  MPI_Alltoallv(in,pnet->inlen,pnet->indisp,type,out,pnet->outlen,
                pnet->outdisp,type,pnet->comm);
  MPI_Type_free(&type);
#else
  for(i=0;i<outsize;i++)
    out[i]=in[i];
#endif
#ifdef DEBUG
  printf("out %d:",pnet->taskid);
  j=0;
  for(i=0;i<outsize;i++)
  {
    if(pnet->outlen[j]+pnet->outdisp[j]==i)
    {
      printf("|");
      j++;
    }
    printf(" %d,%d",out[i].from,out[i].index);
  }
  printf("\n");
#endif
  /* build new connection lists */
  for(i=0;i<outsize;i++)
  {
    rc=pnet_addconnect(pnet,out[i].index,out[i].from);
    if(rc!=PNET_OK)
    {
      free(in);
      free(lo);
      free(hi);
      free(out);
      return rc;
    }
  }
  /* free memory */
  free(in);
  free(lo);
  free(hi);
  free(out);
  return PNET_OK;
} /* of 'pnet_reverse' */
